### ISSUE TITLE
Multiple page resizes during loading

### ISSUE THREAD

**Issue by AlfredMadere** [jkgenser/react-pdf-headless#12]
First off, this is an awesome project!

I noticed some jitter when scrolling to an area allowed the user to watch the pages render, so I throttled the cpu to 20x and noticed that the pages seem to go through 5-6 different size changes each time the user scrolls quickly. Without CPU throttling this just looks like jitter but when slowed down as shown in the attached screenshot its clear whats going on. I'm doing a deep dive right now trying to fix this and will let you know if i'm successful.

https://github.com/user-attachments/assets/696996c8-0a83-4f3e-bbe2-d15c5b9e1a9a

**Comment by AlfredMadere** [comment]
Two of the resizing events are related to these two lines in Page.tsx.

```typescript
          height: `${shouldRender ? 'fit-content' : `${virtualItem.size}px`}`,
          minHeight: `${virtualItem.size-100}px`

```

shouldRender is set to true before the renderPageLayer actually renders which results in the following behavior:

- shouldRender starts as false so the page size is equal to virtualItem.size
- shouldRender is set to true but there is no content in the pageLayer yet so the height is set to the minHeight which is 100px less than the virtualItem.size
- The content is rendered in the pageLayer and because the 'fit-content' class is now applied, the height of the page jumps back up to approximately virtualItem.size

**Comment by jkgenser** [comment]
> Two of the resizing events are related to these two lines in Page.tsx.
> 
> ```ts
>           height: `${shouldRender ? 'fit-content' : `${virtualItem.size}px`}`,
>           minHeight: `${virtualItem.size-100}px`
> ```
> 
> shouldRender is set to true before the renderPageLayer actually renders which results in the following behavior:
> 
> * shouldRender starts as false so the page size is equal to virtualItem.size
> * shouldRender is set to true but there is no content in the pageLayer yet so the height is set to the minHeight which is 100px less than the virtualItem.size
> * The content is rendered in the pageLayer and because the 'fit-content' class is now applied, the height of the page jumps back up to approximately virtualItem.size

@AlfredMadere : Thanks for filing this Issue. 

Appreciate you pointing out some of the potential root causes. Do you have a working patch that works to fix the issue? If not totally cool, I think you provided enough information for me to reproduce on my end - I plan on looking into this!

**Comment by AlfredMadere** [comment]
Hey there, unfortunately I don't yet have a patch.  I'm working on a stripped down version as I don't need rotation or zoom features for my project, If i can get it working there, I will post my findings and or create a PR if the fix is generalizable.

**Comment by AlfredMadere** [comment]
Another thing i'm noticing is that shouldRender toggles on and off twice in each page component every time the scroll stops resulting in the loading component flashing on and off twice. Looking at the profiler it doesn't seem like there are big performance gains from waiting to render until the scrolling stops so i just removed it and it seemed to solve that issue without any real downsides. I also solved the height jittering by just forcing the page-inner-box to always be the same height like so:

```typescript
 height: `${`${virtualItem.size - EXTRA_HEIGHT}px`}`
 ```
 
 And then in order to allow for a loading component that can fill the whole page (maybe centered or something) I made the page take up the full width and height of the page-inner-box. I used tailwind classes but you could translate that to a style tag.
 
 ```jsx
  <div className={'size-full relative'}>
            {renderPageLayer({
              pageIndex: virtualItem.index,
              scale: scale,
              rotate: rotation,
              rotationAdjustment
            })}
          </div>
```
 
Let me know if you want any more info if you decided to fix. Also happy to make a PR in the future when I have a little more time.

**Comment by AlfredMadere** [comment]
Here's what it looks like after those changes at 20x cpu slowdown


https://github.com/user-attachments/assets/3908e582-28d0-4812-a46c-100e620d199a

**Comment by jkgenser** [comment]
That looks awesome! I started with incorporating the height hard coding. I'm thinking through whether to remove the `shouldRender` logic completely. The purpose of that was so that if the scrolling slowed down or was going from only a few pages, then the intermediate pages would render.

But if scrolling from top to bottom for a large document, don't render all the intermediate pages. However, given that the scroll is typically quite quick, maybe it's not worth the extra rendering.

**Comment by AlfredMadere** [comment]
I agree that shouldRender is a good feature that is worth keeping. I removed it in my project because I wasn't sure how to solve the issue of the loading component flashing on and off. Here are two videos, one with, and one without conditional rendering based on shouldRender. You can see how the loading component flashes on and off and at full speed with no CPU throttling it looks like a glitch. Are you experiencing the same thing?


https://github.com/user-attachments/assets/738f3221-ad73-4153-bcf7-3d16fdb4db56


https://github.com/user-attachments/assets/fe46e0b1-095f-457c-88ac-468e2db06c54
